uqcsbot/utils/itee_seminar_utils.py

```diff
@@ -0,0 +1,133 @@
+from uqcsbot import bot
+import requests
+import re
+from dateutil import parser
+from bs4 import BeautifulSoup
+from pytz import timezone
+
+# Utilities for parsing seminar information from the School of ITEE's seminar listing page at
+# https://www.itee.uq.edu.au/seminar-list.
+ITEE_BASE_URL = 'https://www.itee.uq.edu.au'
+ITEE_SEMINAR_LIST_URL = 'https://www.itee.uq.edu.au/seminar-list'
+BRISBANE_TZ = timezone('Australia/Brisbane')
+SEMINAR_DETAILS_REGEX = re.compile('group_seminar_details_element')
+
+
+class InvalidFormatException(Exception):
+    """
+    Raised when an element in a document could not be parsed correctly
+    """
+    def __init__(self, url: str, description: str):
+        self.message = f'{description} on \'{url}\'.'
+        self.url = url
+        super().__init__(self.message, self.url)
+
+
+class HttpException(Exception):
+    """
+    Raised when a HTTP request returns an unsuccessful (i.e. not 200 OK) status
+    code.
+    """
+    def __init__(self, url: str, status_code: int):
+        self.message = f'Received status code {status_code} from \'{url}\'.'
+        self.url = url
+        self.status_code = status_code
+        super().__init__(self.message, self.url, self.status_code)
+
+
+def get_seminars():
+    """
+    Returns summary information for upcoming ITEE seminars, comprising seminar
+    date, seminar title, venue, and an information link.
+    """
+    html = BeautifulSoup(get_seminar_summary_page(), 'html.parser')
+    summary_table = html.find('table', summary='ITEE Seminar List')
+    if (summary_table is None) or (summary_table.tbody is None):
+        # When no seminars are scheduled, no table is shown.
+        return []
+
+    seminar_rows = summary_table.tbody.find_all('tr')
+    seminar_summaries = map(get_seminar_summary, seminar_rows)
+    return list(seminar_summaries)
+
+
+def get_seminar_summary_page():
+    """
+    Returns the content of the page summarising upcoming seminars.
+    :return: The HTML of the page containing upcoming seminar information.
+    """
+    http_response = requests.get(ITEE_SEMINAR_LIST_URL)
+    if http_response.status_code != requests.codes.ok:
+        raise HttpException(ITEE_SEMINAR_LIST_URL, http_response.status_code)
+    return http_response.content
+
+
+def get_seminar_summary(seminar_row):
```

Comment:
Return type annotation. `from typing import Tuple`

I won't add comments for the other issues of this type here.